pkg/response/response.go

```diff
@@ -19,65 +26,243 @@ const (
+	}
+	return buf.String()
+}
+
+func formatYaml(r *Response) string {
+	out, err := yaml.Marshal(r)
+	if err != nil {
+		logger.Error("could not marshal response: error=%q", err)
+	}
 	return string(out)
 }
+
+func formatCsv(r *Response) string {
+	headers := flattenedResponseFields()
+
+	var buf bytes.Buffer
+	writer := csv.NewWriter(&buf)
+
+	err := writer.Write(headers)
+	if err != nil {
+		logger.Error("could not write response: error=%q", err)
+	}
+
+	err = writer.WriteAll(flattenedResponse(r))
+	if err != nil {
+		logger.Error("could not write response: error=%q", err)
+	}
+
+	return buf.String()
+}
+
+// flattenedResponseFields provides a list containing the field labels for a flattened response
+func flattenedResponseFields() []string {
+	return []string{"ID", "REQUEST.ID", "RESULT.ID", "RESULT.KIND", "RESULT.SECRET", "RESULT.MATCH",
+		"RESULT.CONTEXT", "RESULT.ENTROPY", "RESULT.DATE", "RESULT.NOTES", "RESULT.RULE.ID", "RESULT.RULE.DESCRIPTION",
+		"RESULT.RULE.TAGS", "RESULT.CONTACT", "RESULT.LOCATION.VERSION", "RESULT.LOCATION.PATH",
+		"RESULT.LOCATION.RANGE"}
+}
+
+// flattenedResponse takes the response and returns a 2d list of responses in flattenedResponseFields order
+func flattenedResponse(response *Response) [][]string {
+	var flattened [][]string
+
+	for _, result := range response.Results {
+		flattened = append(flattened, []string{
+			sanitizeField(response.ID),
+			sanitizeField(response.RequestID),
+			sanitizeField(result.ID),
+			sanitizeField(result.Kind),
+			sanitizeField(result.Secret),
+			sanitizeField(result.Match),
+			sanitizeField(result.Context),
+			sanitizeField(fmt.Sprintf("%f", result.Entropy)),
+			sanitizeField(result.Date),
+			sanitizeField(flattenMapToString(result.Notes)),
+			sanitizeField(result.Rule.ID),
+			sanitizeField(result.Rule.Description),
+			sanitizeField(strings.Join(result.Rule.Tags, ", ")),
+			sanitizeField(flattenContact(result.Contact)),
+			sanitizeField(result.Location.Version),
+			sanitizeField(result.Location.Path),
+			sanitizeField(fmt.Sprintf("L%dC%d-L%dC%d", result.Location.Start.Line,
+				result.Location.Start.Column, result.Location.End.Line, result.Location.End.Column)),
+		})
+	}
+
+	return flattened
+}
+
+// flattenMapToString creates a flat string with pairs "key: value" only single depth
+func flattenMapToString(m map[string]string) string {
+	var pairs []string
+	for k, v := range m {
+		pairs = append(pairs, fmt.Sprintf("%s: %s", k, v))
+	}
+	return strings.Join(pairs, ", ")
+}
+
+// flattenContact creates a single string with Contact information as "Name <Email>"
+func flattenContact(contact Contact) string {
+	return fmt.Sprintf("%s <%s>", contact.Name, contact.Email)
+}
+
+// sanitizeField takes a string and makes it safe for CSV by replacing newlines and escaping quotes
+// CSV is not a great format for rich/nested data, this makes the data more consistent for the format
+func sanitizeField(value string) string {
```

Comment:
@thewizzy does the CSV writer not handle that automagically? Or is this more for the human one?

Reply:
CSV breaks with nested `'` or newlines. I was shocked at how little escaping it does.